### PR TITLE
[WIP] implementation for basic gaxpy and tests

### PR DESCRIPTION
**Pull request by jieli-matrix**
Todo:

- [ ] Implement `imul!` with
  - [x] Sparse Matrix  - Vector/Matrix
  - [x] Adjoint / Transpose Matrix - Vector/Matrix
- [ ] AD: Generate Jacobian matrix
- [ ] tests
  - [x] Write tests for all implemented methods and unit tests passed
  - [ ] Compared results with ForwardDiff.jl
      
Ask for review:
       Suggestions on mistakes / code-style

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -3,6 +3,12 @@ uuid = "b3c78db8-5351-4c94-92cb-e6a7a02f3abe"
 authors = ["JieLi"]
 version = "0.1.0"
 
+[deps]
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+NiLang = "ab4ef3a6-0b42-11ea-31f6-e34652774712"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
```
Comment (GiggleLiu) [inline Project.toml]: Random is only used in testing. You should put it in the `extra` field. e.g. 
https://github.com/QuantumBFS/YaoArrayRegister.jl/blob/9dd04982359435bca2f93679c23824420229799b/Project.toml#L24

File: test/linalg.jl [hunk 2]
```diff
@@ -0,0 +1,64 @@
+using NiSparseArrays:imul!
+using SparseArrays, Random, Test
+Random.seed!(1234)
+
+# Is declaring approx_rtol here safe? But I don't want to declare approx_rtol in each testset...
+approx_rtol = 100*eps()
+# Would it be reasonable to test accuracy like this? Is there any better readable code?
```
Comment (GiggleLiu) [inline test/linalg.jl]: Looks good to me. Maybe using `const` is better, it will make the code type stable.

```julia
const approx_rtol = 100*eps()
```

File: test/linalg.jl [hunk 3]
```diff
@@ -0,0 +1,64 @@
+using NiSparseArrays:imul!
+using SparseArrays, Random, Test
+Random.seed!(1234)
```
Comment (GiggleLiu) [inline test/linalg.jl]: Set the seed inside each `@testset` rather than here to make the tests having consistent results (reproducible).

**Review by GiggleLiu — COMMENTED** [review]
Looks good in general, cheers!

File: src/linalg.jl [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        @safe size(xA.parent, 1) == size(B, 1) || throw(DimensionMismatch())
+        @safe size(B, 2) == size(C, 2) || throw(DimensionMismatch())
+        if (β != 1, ~)
+            @safe error("only β = 1 is supported, got β = $(β).")
+        end
+        @invcheckoff for k in 1:size(C, 2)
+            @inbounds for col in 1:size(xA.parent, 2)
+                    for j in nzrange(xA.parent, col)
+                    anc2 ← zero(eltype(xA))
+                    anc2 += $t(xA.parent.nzval[j])
```
Comment (GiggleLiu) [inline src/linalg.jl]: I am afraid the backward rule of `y += adjoint(x)` is not defined. The following alternative is prefered
```julia
y += x'
```
For transpose, we do not support it yet (it does not make sense to transpose a scalar). If you think having `adjoint` and `transpose` will make your implementation more consistent, you may add these rules to this [file](https://github.com/GiggleLiu/NiLang.jl/blob/master/src/autodiff/instructs.jl) by submitting a PR.

File: test/linalg.jl (hunk 3 above)
Comment (johnnychen94) [reply test/linalg.jl]: FYI, perhaps surprisingly, I vaguely remember that `@testset` sets its own random seed...

```julia
julia> Random.seed!(1234)
TaskLocalRNG()

julia> @testset begin
           @show rand()
       end;
rand() = 0.7601155742719967

julia> @testset begin
           @show rand()
       end;
rand() = 0.7601155742719967
```

File: Project.toml (hunk 1 above)
Comment (jieli-matrix) [reply Project.toml]: Thanks and I have corrected it in my local workspace~

File: test/linalg.jl (hunk 2 above)
Comment (jieli-matrix) [reply test/linalg.jl]: Thanks and it has been corrected~

File: test/linalg.jl (hunk 3 above)
Comment (jieli-matrix) [reply test/linalg.jl]: Thanks for suggestions from mentors(I haven't considered the impact of random seed in/outside testset on the results before). So maybe there is no need to set the seed in each testset?

File: src/linalg.jl (hunk 4 above)
Comment (jieli-matrix) [reply src/linalg.jl]: Thanks! In the first stage I will only keep the ' (adjoint)'s implementation, I would consider adding transpose when I dive more deep into the reverse computing~

File: test/linalg.jl (hunk 3 above)
Comment (johnnychen94) [reply test/linalg.jl]: There're gripes on this testset-owned random seed behavior, so I would go as @GiggleLiu suggested. But generally, if your algorithm is robust enough, you don't need to care about the random seed value.

File: Project.toml [hunk 5]
```diff
@@ -4,16 +4,17 @@ authors = ["JieLi"]
 version = "0.1.0"
 
 [deps]
+ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
```
Comment (GiggleLiu) [inline Project.toml]: It is not a dependency, please move it to `[extras]` field.

File: src/sparsegrad.jl [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
+    end
+    return J
+end
+
+"""
+    forwarddiff_mv_jacobian(A, x) -> AbstractArray J 
+forwarddiff_mv_jacobian calculates the jacobian with respect to parameter A and x based on multiplication in ForwardDiff
+"""
+function forwarddiff_mv_jacobian(A, x)
```
Comment (GiggleLiu) [inline src/sparsegrad.jl]: This is general multiplication, not just `mv`. This function can also be used to test matrix-matrix multiplication.

File: Project.toml (hunk 5 above)
Comment (jieli-matrix) [reply Project.toml]: yeah...I recognized it when I pushed it to the dev branch...

File: src/sparsegrad.jl (hunk 6 above)
Comment (jieli-matrix) [reply src/sparsegrad.jl]: Right, I'm going to add matrix- matrix multiplication in tests.  May be mm is a better name for the loss?(keep the same routine with pytorch? )

Comment (GiggleLiu) [reply src/sparsegrad.jl]: If you want to change a name, please do it in the following style

```julia
julia> _f(x) = x
_f (generic function with 1 method)

julia> f(x::Vector) = _f(x)
f (generic function with 2 methods)

julia> g(x::Matrix) = _f(x)
g (generic function with 2 methods)
```

One of the golden rule of programming is DRY: don't repeat yourself, it means duplicated input in an interface and code copying are not recommended. If you change a name here, you have to copy the code.

Another comment is, in a test, you can write the code in a more casual style, for example, you do not need to document functions, since these functions will not be exported. Just add some comments to help understanding.
```julia
# calculates the jacobian of multiplication with ForwardDiff
# inputs can be matrices and vectors.
```

The above advices are for maintainability. When you change an implementation, you need to change
* its documentation
* every copy of this code

Most people fail to keep the documentations up to date if the documentation is too detailed, gradually, the software can be not usable. Unless you have a team maintaining your project and clients paying for your software, do not write unnecessary documentation. Note, the above comment remains valid even if your function name and arguments list are changed, unlike a docstring. For the same reason, maintaining multiple copies is impossible.

Comment (jieli-matrix) [reply src/sparsegrad.jl]: Thanks for the delicate guidance! 
Besides the documentation,  I was intended to rename the loss function since in some multiplication cases the type of result array is SparseCSC and we need to implement corresponding loss function in another way (but now I feel such case could be distinguished by type).
``` julia
function mmloss(l, j, y::AbstractArray{<:Complex}, A, x)
function mmloss(l, j, y::AbstractSparseMatrix{<:Complex}, A, x)
```
And `nilang_mm_jacobian` is a general pipeline in calculating jacobian, so would it be better to let the loss be the input argument so as to adapt to more loss cases rather implement the whole jacobian function?

```julia
function nilang_mm_jacobian(A, x::AbstractArray, loss)
    px, pA = params(x), params(A)
    J = zeros(eltype(px), length(px), length(px) + length(pA))
    for j=1:length(px)
        _, _, _, _, gA, gx = NiLang.AD.Grad(loss)(Val(1), 0.0, j, zero(x), A, x)
        J[j,:] = vcat(grad(params(gA)), grad(params(gx)))
    end
    return J
end
```

Comment (johnnychen94) [reply src/sparsegrad.jl]: Weeks ago I complained about the current Nilang AD API isn't very convinient for normal users, maybe it's time to revisit https://github.com/GiggleLiu/NiLang.jl/blob/f24036c478ace467ae71f630e1db2698e38fd64d/src/autodiff/jacobian.jl#L27-L44?

@GiggleLiu maybe open an issue for this gradient/jacobian API and discuss it in a separate thread?

---

@jieli-matrix Don't forget that this package depends on Nilang and not the reverse. We don't need to rebuild the `gradient`/`jacobian` functionality in this package; we should just use whatever NiLang provides, and if that doesn't fit your need or intuition very well, raise an issue in NiLang for improvement.

Comment (johnnychen94) [reply src/sparsegrad.jl]: For this PR, since it has the tendency to be epic, I feel it's perhaps okay to just leave `nilang_mm_jacobian` as it is so that this PR doesn't last forever.

Comment (jieli-matrix) [reply src/sparsegrad.jl]: That's right. We shouldn't implement another jacobian method in this repo, just use the API that NiLang provides.

**Comment by jieli-matrix** [comment]
Methods implemented for Sparse Matrix Multiplication:
- [x] imul!(C::StridedVecOrMat, A::AbstractSparseMatrix, B::DenseInputVecOrMat, α::Number, β::Number)
- [x] imul!(C::StridedVecOrMat, xA::Adjoint{<:Any,<:AbstractSparseMatrix}, B::DenseInputVecOrMat, α::Number, β::Number)
- [x] imul!(C::StridedVecOrMat, X::DenseMatrixUnion, A::AbstractSparseMatrix, α::Number, β::Number)
- [x] imul!(C::StridedVecOrMat, X::Adjoint{<:Any,<:DenseMatrixUnion}, A::AbstractSparseMatrix, α::Number, β::Number)
- [x]  imul!(C::StridedVecOrMat, X::DenseMatrixUnion, xA::Adjoint{<:Any,<:AbstractSparseMatrix}, α::Number, β::Number)
- [ ]  idot(r::T, A::SparseMatrixCSC{T},B::SparseMatrixCSC{T}) where {T}
- [x] idot(r, x::AbstractVector, A::AbstractSparseMatrix{T1}, y::AbstractVector{T2}) where {T1, T2}
- [x] idot(r, x::SparseVector, A::AbstractSparseMatrix{T1}, y::SparseVector{T2}) where {T1, T2}
- [ ] i_spmatmul(A::SparseMatrixCSC{Tv, Ti}, B::SparseMatrixCSC{Tv, Ti}) where {Tv, Ti}

Selected functions have been implemented and passed CI tests. 
Major Contributions in this commits:
- implement idot functions
- revise "three arguments" bad programming style
- write tests in consistent style

I would like to discuss is there performance gap in the implementation of  *idot(r, x::SparseVector, A::AbstractSparseMatrix{T1}, y::SparseVector{T2}) where {T1, T2}* since it differs from the [original implementation](https://github.com/JuliaLang/julia/blob/master/stdlib/SparseArrays/src/linalg.jl#L354). I couldn't find the source code of _spdot (may be it is better double-for-loop in my implementation); besides, may be the order of for loops would influence the performance. (I will try it in my local benchmark tests.)

There are some unexpected errors in idot(r::T, A::SparseMatrixCSC{T},B::SparseMatrixCSC{T}) where {T}, I would go into details in another branch.

**Comment by jieli-matrix** [comment]
Tasks for the next stage:
- [ ] Complete the Sparse Matrix Multiplication implementations
- [ ] Add jacobian tests for CI-passed functions
- [ ]  Generate chain-rules for CI-passed functions
- [ ]  Implement Advanced Sparse Matrix Algorithms
    - lobpcg
    - pca_lowrank
    - svd_lowrank

File: src/NiSparseArrays.jl [hunk 7]
```diff
@@ -1,5 +1,6 @@
 module NiSparseArrays
 
+using LinearAlgebra: eltype
```
Comment (johnnychen94) [inline src/NiSparseArrays.jl:3]: This is just the already exported `eltype` so no need to add this line.

File: src/linalg.jl [hunk 8]
```diff
@@ -92,8 +81,115 @@ end
...
+    end
+    @invcheckoff for col in 1:size(xA.parent, 2)
+        @inbounds for k in nzrange(xA.parent, col)
+                        @routine begin
+                            anc1 ← zero(eltype(xA.parent))
+                            anc1 += (xA.parent.nzval[k])'*α
+                        end
+            @simd for multivec_row in 1:size(X,1)
+                C[multivec_row, xA.parent.rowval[k]] += X[multivec_row, col] * anc1
+            end
```
Comment (johnnychen94) [inline src/linalg.jl:110]: I didn't run it so just a question. Does this `@simd` annotation work here? Say, does it bring performance differences?

File: src/linalg.jl [hunk 9]
```diff
@@ -92,8 +81,115 @@ end
...
+            ~@routine
+        end
+    end
+end
+
+@i function idot(r::T, A::SparseMatrixCSC{T},B::SparseMatrixCSC{T}) where {T}
+    @routine @invcheckoff begin
+        (m, n) ← size(A)
```
Comment (johnnychen94) [inline src/linalg.jl:118]: We don't need this. One can directly use `size(A, 1)` and `size(A, 2)` later.

File: src/linalg.jl [hunk 10]
```diff
@@ -92,8 +81,115 @@ end
...
+    @invcheckoff @inbounds for j = 1:n
+        @routine begin
+            ia1 ← A.colptr[j]
+            ib1 ← B.colptr[j]
+            ia2 ← A.colptr[j+1]
+            ib2 ← B.colptr[j+1]
+            ia ← ia1
+            ib ← ib1
+        end
+        @inbounds for i=1:ia2-ia1+ib2-ib1-1
```
Comment (johnnychen94) [inline src/linalg.jl:131]: @GiggleLiu will this introduce implicit memory allocation?

File: src/linalg.jl [hunk 11]
```diff
@@ -92,8 +81,115 @@ end
...
+            ~@routine 
+        end
+    end
+end 
+
+@i function idot(r, x::SparseVector, A::AbstractSparseMatrix{T1}, y::SparseVector{T2}) where {T1, T2}
+    @safe length(x) == size(A, 1) || throw(DimensionMismatch())
+    @safe length(y) == size(A, 2) || throw(DimensionMismatch())
+
+    @invcheckoff for (yi, yv) in zip(y.nzind, y.nzval)
```
Comment (johnnychen94) [inline src/linalg.jl:181]: Some `@inbounds` annotation maybe?

**Review by johnnychen94 — COMMENTED** [review]
Just point out some places that look suspicious to me. I didn't read the implementation details (because I don't know), so It's likely that I'm wrong here.

Need to remove the personal `.vscode/launch.json` from commit history.

File: src/linalg.jl (hunk 8 above)
Comment (GiggleLiu) [reply src/linalg.jl:110]: It is allowed here, whether it brings speed up requires benchmarking.

File: src/linalg.jl (hunk 10 above)
Comment (GiggleLiu) [reply src/linalg.jl:131]: No, its inverse is
```julia
var = ia2-ia1+ib2-ib1-1
@inbounds for i=ia2-ia1+ib2-ib1-1:-1:1
    ...
end
@assert var == ia2-ia1+ib2-ib1-1
```
The `var` is "allocated" on register. In addition, if you put a `@invcheckoff` before this statement, the `@assert` statement will be removed.

File: src/linalg.jl (hunk 11 above)
Comment (GiggleLiu) [reply src/linalg.jl:181]: `@inbounds` would help improving the performance.

File: src/NiSparseArrays.jl (hunk 7 above)
Comment (jieli-matrix) [reply src/NiSparseArrays.jl:3]: > using LinearAlgebra: eltype

It was added by VSCode and sometimes I just forgot to delete it...
> .vscode/launch.json removed

Maybe I could add filter in .gitignore file?

Comment (johnnychen94) [reply src/NiSparseArrays.jl:3]: Yes it makes sense to ignore the entire .vscode folder in gitignore

**Comment by johnnychen94** [comment]
This PR has become chaotic because it is involved with too many independent features and closely related PRs (e.g., #5 #7 ). For this reason, I've split this PR into a smaller one #8 from 85459c3, we should resolve https://github.com/jieli-matrix/NiSparseArrays.jl/pull/2#commitcomment-53866423 first and polish that PR first.

We should keep this branch untouched until #8 is merged.

I say these features are independent because it's totally doable to add only one method at one time (PR).

**Comment by GiggleLiu** [comment]
Close for having no updates, feel free to check out this branch, merge master, keep working on it and submit a new PR.
